asm/amd/regs_state.go

```diff
@@ -3,64 +3,310 @@
-		return 11
-	case x86asm.R12, x86asm.R12L:
-		return 12
-	case x86asm.R13, x86asm.R13L:
-		return 13
-	case x86asm.R14, x86asm.R14L:
-		return 14
-	case x86asm.R15, x86asm.R15L:
-		return 15
-	case x86asm.RSP, x86asm.ESP:
-		return 16
-	case x86asm.RIP:
-		return 17
-	default:
-		return 0
+	idx, _ := regIndexWithSize(reg)
+	return idx
+}
+
+func regIndexWithSize(reg x86asm.Reg) (idx, size int) {
+	if reg > 0 && int(reg) < len(regIndexTable) {
+		e := regIndexTable[reg]
+		return e.idx, e.sz
 	}
+	return 0, 0
 }
 
 type RegsState struct {
-	regs [18]regState
+	regs [18]variable.U64
 }
 
-func (r *RegsState) Set(reg x86asm.Reg, value, loadedFrom uint64) {
-	r.regs[regIndex(reg)].Value = value
-	r.regs[regIndex(reg)].LoadedFrom = loadedFrom
+func (r *RegsState) Set(reg x86asm.Reg, v variable.U64) {
+	idx, sz := regIndexWithSize(reg)
+	if sz != size64 {
+		v = variable.Crop(v, sz)
+	}
+	if debugPrinting {
+		if reg != x86asm.RIP {
+			fmt.Printf("                               -> | %6s = %s\n", reg, v.String())
+		}
+	}
+	r.regs[idx] = v
+}
+
+func (r *RegsState) Get(reg x86asm.Reg) variable.U64 {
+	idx, sz := regIndexWithSize(reg)
+	res := r.regs[idx]
+	if sz != size64 {
+		res = variable.Crop(res, sz)
+	}
+	return res
 }
 
-func (r *RegsState) Get(reg x86asm.Reg) (value, loadedFrom uint64) {
-	return r.regs[regIndex(reg)].Value, r.regs[regIndex(reg)].LoadedFrom
+type Interpreter struct {
+	Regs        RegsState
+	code        []byte
+	CodeAddress variable.U64
+	pc          int
 }
 
-type regState struct {
-	LoadedFrom uint64
-	Value      uint64
+func NewInterpreter(code []byte) Interpreter {
+	it := Interpreter{code: code, CodeAddress: variable.Var("code address")}
+	it.initRegs()
+	return it
+}
+
+func (i *Interpreter) initRegs() {
+	i.Regs.regs[0] = variable.Var("invali reg")
+	i.Regs.regs[regIndex(x86asm.RAX)] = variable.Var("initial RAX")
+	i.Regs.regs[regIndex(x86asm.RCX)] = variable.Var("initial RCX")
+	i.Regs.regs[regIndex(x86asm.RDX)] = variable.Var("initial RDX")
+	i.Regs.regs[regIndex(x86asm.RBX)] = variable.Var("initial RBX")
+	i.Regs.regs[regIndex(x86asm.RSP)] = variable.Var("initial RSP")
+	i.Regs.regs[regIndex(x86asm.RBP)] = variable.Var("initial RBP")
+	i.Regs.regs[regIndex(x86asm.RSI)] = variable.Var("initial RSI")
+	i.Regs.regs[regIndex(x86asm.RDI)] = variable.Var("initial RDI")
+	i.Regs.regs[regIndex(x86asm.R8)] = variable.Var("initial R8")
+	i.Regs.regs[regIndex(x86asm.R9)] = variable.Var("initial R9")
+	i.Regs.regs[regIndex(x86asm.R10)] = variable.Var("initial R10")
+	i.Regs.regs[regIndex(x86asm.R11)] = variable.Var("initial R11")
+	i.Regs.regs[regIndex(x86asm.R12)] = variable.Var("initial R12")
+	i.Regs.regs[regIndex(x86asm.R13)] = variable.Var("initial R13")
+	i.Regs.regs[regIndex(x86asm.R14)] = variable.Var("initial R14")
+	i.Regs.regs[regIndex(x86asm.R15)] = variable.Var("initial R15")
+	i.Regs.regs[regIndex(x86asm.RIP)] = variable.Var("initial RIP")
+}
+
+func (i *Interpreter) Loop() error {
+	return i.LoopWithBreak(nil)
```

Comment:
```suggestion
	return i.LoopWithBreak(function(x86asm.Inst) { return false })
```
to avoid the `breakLoop != nil` test below